controllers/datamovement_controller.go

```diff
@@ -86,314 +83,237 @@ func (r *DataMovementReconciler) Reconcile(ctx context.Context, req ctrl.Request
-				status,
-				message,
-			)
+		return r.Status().Update(ctx, dm)
+	})
 
-			now := metav1.NowMicro()
-			dm.Status.EndTime = &now
-		}
+	if errors.IsConflict(err) {
 
-		if err := r.Status().Update(ctx, dm); err != nil {
-			log.Error(err, "Failed to transition to finished state")
-			return ctrl.Result{}, err
-		}
+		log.Info("Status update conflicted")
 
-		return ctrl.Result{}, nil
+		// Record the completion internally so when we requeue we won't
+		// restart the operation but update the resource with saved status.
+		r.completions.Store(req.Name, dm.Status)
 
-	case nnfv1alpha1.DataMovementConditionTypeFinished:
-		return ctrl.Result{}, nil // Already finished, do nothing further.
+		return ctrl.Result{Requeue: true}, nil
 	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *DataMovementReconciler) validateSpec(dm *nnfv1alpha1.NnfDataMovement) error {
+// Retrieve the NNF Nodes that are the target of the data movement operation
+func (r *DataMovementReconciler) getStorageNodeNames(ctx context.Context, dm *nnfv1alpha1.NnfDataMovement) ([]string, error) {
 
-	// Permit empty data movement resources; this results in a data movement resource
-	// that monitors existing rsync resources
-	if dm.Spec.Source == nil && dm.Spec.Destination == nil {
-		return nil
+	// If this is a node data movement request (not in the default/dm namespaces) then the target NNF Node is ourselves.
+	if dm.Namespace != metav1.NamespaceDefault && dm.Namespace != dmv1alpha1.DataMovementNamespace {
+		return []string{"localhost"}, nil
 	}
 
-	if dm.Spec.Source == nil || len(dm.Spec.Source.Path) == 0 {
-		return fmt.Errorf("source path must be defined")
-	}
-	if dm.Spec.Destination == nil || len(dm.Spec.Destination.Path) == 0 {
-		return fmt.Errorf("destination path must be defined")
+	// Otherwise, this is a system wide data movement request we target the NNF Nodes that are defined in the storage specification
+	var storageRef *corev1.ObjectReference
+	if dm.Spec.Source.Storage.Kind == reflect.TypeOf(nnfv1alpha1.NnfStorage{}).Name() {
+		storageRef = dm.Spec.Source.Storage
+	} else if dm.Spec.Destination.Storage.Kind == reflect.TypeOf(nnfv1alpha1.NnfStorage{}).Name() {
+		storageRef = dm.Spec.Destination.Storage
+	} else {
+		return nil, fmt.Errorf("Neither source or destination is of NNF Storage type")
 	}
 
-	// If destination is just "path" this must be a lustre file system
-	if dm.Spec.Source.Storage == nil && dm.Spec.Destination.Storage == nil {
-		return fmt.Errorf("one of source or destination must be a storage instance")
+	storage := &nnfv1alpha1.NnfStorage{}
+	if err := r.Get(ctx, types.NamespacedName{Name: storageRef.Name, Namespace: storageRef.Namespace}, storage); err != nil {
+		return nil, err
 	}
 
-	return nil
-}
-
-func (r *DataMovementReconciler) isLustre2Lustre(ctx context.Context, dm *nnfv1alpha1.NnfDataMovement) (isLustre bool, err error) {
-	if dm.Spec.Source == nil || dm.Spec.Destination == nil {
-		return false, nil
+	targetAllocationSetIndex := -1
```

Comment:
Confirm storage.Spec.FilesystemType=lustre before  going down here.  This demand for an OST is kind of abrupt-ish.